UnityProject/Assets/Scripts/Player/IPlayerPossessable.cs

```diff
@@ -1,19 +1,185 @@
 using System.Collections.Generic;
+using Core.Utils;
 using HealthV2;
+using Messages.Server;
+using Mirror;
+using Systems.Ai;
 using UnityEngine;
 
 public interface IPlayerPossessable
 {
 
-	LivingHealthMasterBase LHMB { get; set; }
+	public GameObject GameObject { get; }
+	public IPlayerPossessable Possessing { get; set; }
 
-	virtual LivingHealthMasterBase LivingHealthMasterBase()
+	public GameObject PossessingObject { get; set; }
+
+	public Mind PossessingMind { get; set; }
+
+	public IPlayerPossessable PossessedBy { get; set; }
+
+	public MindNIPossessingEvent OnPossessedBy   { get; set; }
+
+	public void InternalOnEnterPlayerControl(GameObject PreviouslyControlling, Mind mind, bool IsServer)
+	{
+		if (IsServer)
+		{
+			//can observe their new inventory
+			var dynamicItemStorage = GameObject.GetComponent<DynamicItemStorage>();
+			if (dynamicItemStorage != null)
+			{
+				dynamicItemStorage.ServerAddObserverPlayer(GameObject);
+				PlayerPopulateInventoryUIMessage.Send(dynamicItemStorage,
+					GameObject); //TODO should we be using the players body as game object???
+			}
+			// If the player is inside a container, send a ClosetHandlerMessage.
+			// The ClosetHandlerMessage will attach the container to the transfered player.
+			var playerObjectBehavior = GameObject.GetComponent<UniversalObjectPhysics>();
+			if (playerObjectBehavior != null )
+			{
+				FollowCameraMessage.Send(GameObject, playerObjectBehavior.GetRootObject);
+			}
+
+			PossessAndUnpossessMessage.Send(GameObject, GameObject, PreviouslyControlling);
+
+
+			var PS = GameObject.GetComponent<PlayerScript>();
+			if (PS)
+			{
+				PS.SetMind(mind); //TODO unset
+			}
+
+
+			var health = GameObject.GetComponent<LivingHealthMasterBase>();
+			if (health != null)
+			{
+				mind.bodyMobID = health.mobID;
+			}
+
+
+
+			var transfers = GameObject.GetComponents<IOnPlayerTransfer>();
+
+			foreach (var transfer in transfers)
+			{
+				transfer.OnPlayerTransfer(mind.ControlledBy);
+			}
+		}
+
+		if (GameObject.GetComponent<NetworkIdentity>().hasAuthority)
+		{
+			UIManager.Display.RejoinedEvent();
+			IPlayerControllable input = GameObject.GetComponent<IPlayerControllable>();
+
+			if (GameObject.TryGetComponent<AiMouseInputController>(out var aiMouseInputController))
+			{
+				input = aiMouseInputController;
+			}
+
+			PlayerManager.SetPlayerForControl(GameObject, input);
+			var dynamicItemStorage = GameObject.GetComponent<DynamicItemStorage>();
+			if (dynamicItemStorage != null)
+			{
+				dynamicItemStorage.UpdateSlots(	dynamicItemStorage.GetSetData, 	dynamicItemStorage.GetSetData);
+			}
+		}
+
+		OnEnterPlayerControl( PreviouslyControlling,  mind,  IsServer);
+	}
+
+	public void OnEnterPlayerControl(GameObject PreviouslyControlling, Mind mind, bool IsServer);
+
+	public bool IsRelatedToObject(GameObject Object)
+	{
+		if (PossessingObject == Object)
+		{
+			return true;
+		}
+
+		if (Possessing != null && Possessing.IsRelatedToObject(Object))
```

Comment:
can be merged into the return statement below

Reply:
I think it makes it easier to reach a separate this stuff out

Reply:
easier to read if it separated out, and also adding stuff

Reply:
I don't see any added stuff, so it's cleaner if it returns immediately